🎨 Add examples to pydantic response models

bisgaard-itis opened the pull request:
<!-- Title Annotations:

  WIP: work in progress
  🐛    Fix a bug.
  ✨    Introduce new features.
  🎨    Enhance existing feature.
  ♻️    Refactor code.
  🚑️    Critical hotfix.
  ⚗️    Perform experiments.
  ⬆️    Upgrade dependencies.
  📝    Add or update documentation.
  🔨    Add or update development scripts.
  🔒️    Fix security issues.
  ⚠️    Changes in ops configuration etc. are required before deploying.
        [ Please add a link to the associated ops-issue or PR, such as in https://github.com/ITISFoundation/osparc-ops-environments or https://git.speag.com/oSparc/osparc-infra ]
  🗃️    Database table changed (relevant for devops).
  👽️    Public API changes (meaning: dev features are moved to being exposed in production)
  🚨    Do manual testing when deployed

or from https://gitmoji.dev/
-->

## What do these changes do?
- Ensure that the api of the api-server is as backwards compatible as is possible. This means the `openapi.json` does not break compatibility except in the following cases:
  - Nullable fields in our pydantic models will not be nullable in the spec (this is not actually an api change but just a change in the `openapi.json`)
  - For the paginated endpoints, the response content (for `status=200`) makes certain fields required but nullable. I was not able to restore backwards compatibility because the pagination is handled internally in https://github.com/uriyyo/fastapi-pagination.
- Fixed several api-server unit tests which were failing after updates to pydantic models (v1 -> v2) in other services.
- For testing purposes on the `osparc` python client side we need to have examples attached to all response models in the api-server. 
- This PR adds examples to `JobMetadata` and `JobMetadataUpdate`


## Related issue/s

<!-- Link pull request to an issue
  SEE https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

- resolves ITISFoundation/osparc-issues#428
- fixes #26

  If openapi changes are provided, optionally point to the swagger editor with new changes
  Example [openapi.json specs](https://editor.swagger.io/?url=https://raw.githubusercontent.com/<github-username>/osparc-simcore/is1133/create-api-for-creation-of-pricing-plan/services/web/server/src/simcore_service_webserver/api/v0/openapi.yaml)
-->
- https://github.com/ITISFoundation/osparc-simcore-clients/pull/202

## How to test

<!-- Give REVIEWERS some hits or code snippets on how could this be tested -->

## Dev-ops checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

<!-- Some checks that might help your code run stable on production, and help devops assess criticality.
Modified from https://oschvr.com/posts/what-id-like-as-sre/

- How can DevOps check the health of the service ?
- How can DevOps safely and gracefully restart the service ?
- How and why would this code fail ?
- What kind of metrics are you exposing ?
- Is there any documentation/design specification for the service ?
- How (e.g. through which loglines) can DevOps detect unexpected situations that require escalation to human ?
- What are the resource limitations (CPU, RAM) expected for this service ?
- Are all relevant variables documented and adjustable via environment variables (i.e. no hardcoded magic numbers) ?
-->

giancarloromeo commented on services/api-server/src/simcore_service_api_server/models/schemas/jobs.py:
```diff
@@ -141,6 +153,20 @@ class JobMetadata(BaseModel):
+                },
+                "url": "https://f02b2452-1dd8-4882-b673-af06373b41b3.fake",
+            }
+        }
```
```suggestion
    model_config = ConfigDict(
        json_schema_extra = {
            "example": {
                "job_id": "3497e4de-0e69-41fb-b08f-7f3875a1ac4b",
                "metadata": {
                    "bool": "true",
                    "int": "42",
                    "float": "3.14",
                    "str": "hej med dig",
                },
                "url": "https://f02b2452-1dd8-4882-b673-af06373b41b3.fake",
            }
        }
    )
```

bisgaard-itis replied:
Yes, thanks I already updated

giancarloromeo commented on services/api-server/src/simcore_service_api_server/models/schemas/jobs.py:
```diff
@@ -133,6 +133,18 @@ class JobMetadataUpdate(BaseModel):
+                    "str": "hej med dig",
+                }
+            }
+        }
```
```suggestion
    model_config = ConfigDict(
        json_schema_extra = {
            "example": {
                "metadata": {
                    "bool": "true",
                    "int": "42",
                    "float": "3.14",
                    "str": "hej med dig",
                }
            }
        }
    )
```

bisgaard-itis replied:
Same here

giancarloromeo reviewed (approved):
Thank you! Just adapted Config to the new Pydantic v2 way.

pcrespov reviewed (approved):
thx!

@bisgaard-itis  once this PR is merged in master, please try to resync this commit in [pydantic_v2_migration_do_not_squash_updates](https://github.com/ITISFoundation/osparc-simcore/tree/pydantic_v2_migration_do_not_squash_updates) branch. This will help us all in getting pydantic upgraded ASAP. thx

mrnicegyu11 commented on services/api-server/openapi.json:
```diff
@@ -5934,7 +5934,17 @@
+            "bool": "true",
+            "float": "3.14",
+            "int": "42",
+            "str": "hej med dig"
```
no klingon language examples please

mrnicegyu11 reviewed (commented):
tldr & good luck < 3

sanderegg reviewed (approved):
thanks

bisgaard-itis commented:
I have checked (using https://github.com/Tufin/oasdiff) that the api-server's `openapi.json` on commit  https://raw.githubusercontent.com/bisgaard-itis/osparc-simcore/6c582d8ebe687dbde87e81e1bbf5da30154b663c/services/api-server/openapi.json of this branch and on master's commit https://raw.githubusercontent.com/ITISFoundation/osparc-simcore/4abb9f23e99c5cd854e348a81f8e8e4e4c88c691/services/api-server/openapi.json are compatible except for the paginated endpoints (as explained above):

### New Endpoints: None
-----------------------

### Deleted Endpoints: None
---------------------------

### Modified Endpoints: 3
-------------------------
GET /v0/files:search
- Responses changed
  - Modified response: 200
    - Content changed
      - Modified media type: application/json
        - Schema changed
          - Required changed
            - New required property: limit
            - New required property: offset
            - New required property: total
          - Properties changed
            - Modified property: limit
              - Property 'AnyOf' changed
                - Schemas added: subschema #1, subschema #2
              - Type changed from 'integer' to ''
              - Min changed from 1 to null
            - Modified property: links
              - Required changed
                - New required property: first
                - New required property: last
                - New required property: next
                - New required property: prev
                - New required property: self
              - Properties changed
                - Modified property: first
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: last
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: next
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: prev
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: self
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
            - Modified property: offset
              - Property 'AnyOf' changed
                - Schemas added: subschema #1, subschema #2
              - Type changed from 'integer' to ''
              - Min changed from 0 to null
            - Modified property: total
              - Property 'AnyOf' changed
                - Schemas added: subschema #1, subschema #2
              - Type changed from 'integer' to ''
              - Min changed from 0 to null

GET /v0/solvers/{solver_key}/releases/{version}/jobs/page
- Responses changed
  - Modified response: 200
    - Content changed
      - Modified media type: application/json
        - Schema changed
          - Required changed
            - New required property: limit
            - New required property: offset
            - New required property: total
          - Properties changed
            - Modified property: limit
              - Property 'AnyOf' changed
                - Schemas added: subschema #1, subschema #2
              - Type changed from 'integer' to ''
              - Min changed from 1 to null
            - Modified property: links
              - Required changed
                - New required property: first
                - New required property: last
                - New required property: next
                - New required property: prev
                - New required property: self
              - Properties changed
                - Modified property: first
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: last
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: next
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: prev
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: self
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
            - Modified property: offset
              - Property 'AnyOf' changed
                - Schemas added: subschema #1, subschema #2
              - Type changed from 'integer' to ''
              - Min changed from 0 to null
            - Modified property: total
              - Property 'AnyOf' changed
                - Schemas added: subschema #1, subschema #2
              - Type changed from 'integer' to ''
              - Min changed from 0 to null

GET /v0/studies
- Responses changed
  - Modified response: 200
    - Content changed
      - Modified media type: application/json
        - Schema changed
          - Required changed
            - New required property: limit
            - New required property: offset
            - New required property: total
          - Properties changed
            - Modified property: limit
              - Property 'AnyOf' changed
                - Schemas added: subschema #1, subschema #2
              - Type changed from 'integer' to ''
              - Min changed from 1 to null
            - Modified property: links
              - Required changed
                - New required property: first
                - New required property: last
                - New required property: next
                - New required property: prev
                - New required property: self
              - Properties changed
                - Modified property: first
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: last
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: next
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: prev
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
                - Modified property: self
                  - Property 'AnyOf' changed
                    - Schemas added: subschema #1, subschema #2
                  - Type changed from 'string' to ''
            - Modified property: offset
              - Property 'AnyOf' changed
                - Schemas added: subschema #1, subschema #2
              - Type changed from 'integer' to ''
              - Min changed from 0 to null
            - Modified property: total
              - Property 'AnyOf' changed
                - Schemas added: subschema #1, subschema #2
              - Type changed from 'integer' to ''
              - Min changed from 0 to null

giancarloromeo reviewed (commented):
Already merged, superthanks :ok_hand:
